File: runtimes/elixir/test/security_limits_test.exs
defmodule Twilic.SecurityLimitsTest do
  use ExUnit.Case
  alias Twilic.Wire
  alias Twilic.Errors.TwilicError

  test "budget persists through byte reads" do
    reader = Wire.new_reader(<<0>>)
    reader = Wire.claim_output(reader, 100)
    {_, reader} = Wire.read_u8(reader)
    assert_raise TwilicError, fn -> Wire.claim_output(reader, 100) end
  end

  test "depth and invalid lengths" do
    assert_raise TwilicError, fn -> Twilic.V2.decode(:binary.copy(<<0xA1>>, 70) <> <<0xC0>>) end
    assert_raise TwilicError, fn -> Wire.read_exact(-1, Wire.new_reader(<<0>>)) end
  end
end

File: runtimes/ruby/test/security_limits_test.rb
require "test_helper"
class SecurityLimitsTest < Minitest::Test
  def test_depth_and_budget
    assert_raises(Twilic::Core::Errors::TwilicError) { Twilic.decode(([0xa1] * 70 + [0xc0]).pack("C*")) }
    reader = Twilic::Core::Wire::Reader.new("\0")
    reader.claim_output(100)
    assert_raises(Twilic::Core::Errors::TwilicError) { reader.claim_output(100) }
    assert_raises(Twilic::Core::Errors::TwilicError) { reader.read_exact(-1) }
  end
end
